actors/builtin/reward/reward_state.go

```diff
@@ -3,305 +3,68 @@ package reward
-	// `denominator` is the accumulator for denominators of the series terms.
-	denominator := denominatorBase
-
-	// `ret` is an _additive_ accumulator for partial sums of the series, and
-	// carries a _fixed-point_ representation rather than a rational
-	// representation. This just means it has an implicit denominator of
-	// 2^(FixedPoint).
-	ret := big.Zero()
-
-	// The first term computed has order 1; the final term has order 24.
-	for n := int64(1); n < int64(25); n++ {
-
-		// Multiplying the denominator by `n` on every loop accounts for the
-		// `n!` (factorial) in the denominator of the series.
-		denominator = big.Mul(denominator, big.NewInt(n))
-
-		// Left-shift and divide to convert rational into fixed-point.
-		term := big.Div(big.Lsh(numerator, MintingOutputFixedPoint), denominator)
-
-		// Accumulate the fixed-point result into the return accumulator.
-		ret = big.Add(ret, term)
-
-		// Multiply the rational representation of (-λt) into the term accumulators
-		// for the next iteration.  Doing this here in the loop allows us to save a
-		// couple bigint operations by initializing numerator and denominator
-		// directly instead of multiplying by 1.
-		numerator = big.Mul(numerator, numeratorBase)
-		denominator = big.Mul(denominator, denominatorBase)
-
-		// If the denominator has grown beyond the necessary precision, then we can
-		// truncate it by right-shifting. As long as we right-shift the numerator
-		// by the same number of bits, all we have done is lose unnecessary
-		// precision that would slow down the next iteration's multiplies.
-		denominatorLen := big.BitLen(denominator)
-		unnecessaryBits := denominatorLen - MintingOutputFixedPoint
-
-		numerator = big.Rsh(numerator, unnecessaryBits)
-		denominator = big.Rsh(denominator, unnecessaryBits)
-
+	for st.CumsumRealized.GreaterThan(st.CumsumBaseline) {
+		st.EffectiveNetworkTime++
+		st.CumsumBaseline = big.Add(st.CumsumBaseline, BaselinePowerAt(st.EffectiveNetworkTime))
 	}
-
-	return ret
 }
 
-// Minting Function Wrapper
-//
-// Intent
-//   The necessary calling conventions for the function above are unwieldy:
-//   the common case is to supply the canonical Lambda, multiply by some other
-//   number, and right-shift down by MintingOutputFixedPoint. This convenience
-//   wrapper implements those conventions. However, it does NOT implement
-//   left-shifting the input by the MintingInputFixedPoint, because baseline
-//   minting will actually supply a fractional input.
-func mintingFunction(factor big.Int, t big.Int) big.Int {
-	return big.Rsh(big.Mul(factor, taylorSeriesExpansion(LambdaNum, LambdaDen, t)), MintingOutputFixedPoint)
+// Takes in a current realized power for a reward epoch and computes
+// and updates reward state to track reward for the next epoch
+func (st *State) updateToNextEpochWithReward(currRealizedPower abi.StoragePower) {
+	prevRewardTheta := computeRTheta(st.EffectiveNetworkTime, st.CumsumRealized, st.CumsumBaseline)
```

Comment:
We update the 𝜃 to be 𝜃(t-1) over here:
https://github.com/filecoin-project/specs-actors/pull/617/files#diff-2e9214f532a06a983c77ad9390267da6R119-R122
which calls https://github.com/filecoin-project/specs-actors/pull/617/files#diff-15478070063dd16268193fab3a001125R50

So no matter number of null rounds the `st.EffectiveNetworkTime, st.CumsumRealized, st.CumsumBaseline` correspond to  `𝜃(t-1)`.

Reply:
Cool, that’s exactly the sort of thing I was missing 😄

My one remaining concern is that it looks like that catch-up loop will continue to update until `st.Epoch` == `rt.CurrEpoch`, which seems to indicate that after doing the `updateToNextEpochWithReward`, `st.Epoch` will be `rt.CurrEpoch + 1` - is that one too many updates? Or is `rt.CurrEpoch` what I think of conceptually as "t-1"? Put another way, if there aren't any null rounds, are we confident that the catch-up loop will run 0 times rather than once?

Reply:
You can think of `st.Epoch` as "for which epoch" `st.ThisEpochReward` is valid. This is the epoch after the computation is performed, as the computation is performed at the epoch end (cron tick).

Reply:
Perfect ☺️